## Project Milestone 2

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
# Open dataset from drive called fraudTrain.csv

data = pd.read_csv('D:/Bellevue Class documents/DSC 410/fraudTrain.csv')
data

Unnamed: 0 trans_date_trans_time               cc_num  \
0                 0   2019-01-01 00:00:18     2703186189652095   
1                 1   2019-01-01 00:00:44         630423337322   
2                 2   2019-01-01 00:00:51       38859492057661   
3                 3   2019-01-01 00:01:16     3534093764340240   
4                 4   2019-01-01 00:03:06      375534208663984   
...             ...                   ...                  ...   
1296670     1296670   2020-06-21 12:12:08       30263540414123   
1296671     1296671   2020-06-21 12:12:19     6011149206456997   
1296672     1296672   2020-06-21 12:12:32     3514865930894695   
1296673     1296673   2020-06-21 12:13:36     2720012583106919   
1296674     1296674   2020-06-21 12:13:37  4292902571056973207   

                                    merchant       category     amt  \
0                 fraud_Rippin, Kub and Mann       misc_net    4.97   
1            fraud_Heller, Gutmann and Zieme    grocery_pos  107.23   
2                       fraud_Lind-Buckridge  entertainment  220.11   
3         fraud_Kutch, Hermiston and Farrell  gas_transport   45.00   
4                        fraud_Keeling-Crist       misc_pos   41.96   
...                                      ...            ...     ...   
1296670                    fraud_Reichel Inc  entertainment   15.56   
1296671             fraud_Abernathy and Sons    food_dining   51.70   
1296672                 fraud_Stiedemann Ltd    food_dining  105.93   
1296673  fraud_Reinger, Weissnat and Strosin    food_dining   74.90   
1296674  fraud_Langosh, Wintheiser and Hyatt    food_dining    4.30   

               first       last gender                         street  ...  \
0           Jennifer      Banks      F                 561 Perry Cove  ...   
1          Stephanie       Gill      F   43039 Riley Greens Suite 393  ...   
2             Edward    Sanchez      M       594 White Dale Suite 530  ...   
3             Jeremy      White      M    9443 Cynthia Court Apt. 038  ...   
4              Tyler     Garcia      M               408 Bradley Rest  ...   
...              ...        ...    ...                            ...  ...   
1296670         Erik  Patterson      M       162 Jessica Row Apt. 072  ...   
1296671      Jeffrey      White      M  8617 Holmes Terrace Suite 651  ...   
1296672  Christopher  Castaneda      M     1632 Cohen Drive Suite 639  ...   
1296673       Joseph     Murray      M           42933 Ryan Underpass  ...   
1296674      Jeffrey      Smith      M           135 Joseph Mountains  ...   

             lat      long  city_pop                                job  \
0        36.0788  -81.1781      3495          Psychologist, counselling   
1        48.8878 -118.2105       149  Special educational needs teacher   
2        42.1808 -112.2620      4154        Nature conservation officer   
3        46.2306 -112.1138      1939                    Patent attorney   
4        38.4207  -79.4629        99     Dance movement psychotherapist   
...          ...       ...       ...                                ...   
1296670  37.7175 -112.4777       258                       Geoscientist   
1296671  39.2667  -77.5101       100   Production assistant, television   
1296672  32.9396 -105.8189       899                    Naval architect   
1296673  43.3526 -102.5411      1126              Volunteer coordinator   
1296674  45.8433 -113.8748       218           Therapist, horticultural   

                dob                         trans_num   unix_time  merch_lat  \
0        1988-03-09  0b242abb623afc578575680df30655b9  1325376018  36.011293   
1        1978-06-21  1f76529f8574734946361c461b024d99  1325376044  49.159047   
2        1962-01-19  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704   
3        1967-01-12  6b849c168bdad6f867558c3793159a81  1325376076  47.034331   
4        1986-03-28  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999   
...             ...                          

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1296675 entries, 0 to 1296674
Data columns (total 23 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   Unnamed: 0             1296675 non-null  int64  
 1   trans_date_trans_time  1296675 non-null  object 
 2   cc_num                 1296675 non-null  int64  
 3   merchant               1296675 non-null  object 
 4   category               1296675 non-null  object 
 5   amt                    1296675 non-null  float64
 6   first                  1296675 non-null  object 
 7   last                   1296675 non-null  object 
 8   gender                 1296675 non-null  object 
 9   street                 1296675 non-null  object 
 10  city                   1296675 non-null  object 
 11  state                  1296675 non-null  object 
 12  zip                    1296675 non-null  int64  
 13  lat                    1296675 non-null  float64
 14  long              

# Milestone 3 Feature Engineering

I changed datasets to one that had named columns and generated information due to the confidential nature of the data.

### Chosing features

In [6]:
data.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   

      last gender                        street  ...      lat      long  \
0    Banks      F                561 Perry Cove  ...  36.0788  -81.1781   
1     Gill      F  43039 Riley Greens Suite 393  ...  48.8878 -118.2105   
2  Sanchez      M      594 White Dale Suite 530  ...  42.1808 -112.2620   
3    White      M   9443 Cynthia Court Apt. 038  ...  46.2306 -112.1138   
4   Garcia      M              408 Bradley Rest  ...  38.4207  -79.4629   

   city_pop                                job         dob  \
0      3495          Psychologist, counselling  1988-03-09   
1       149  Special educational needs teacher  1978-06-21   
2      4154        Nature conservation officer  1962-01-19   
3      1939                    Patent attorney  1967-01-12   
4        99     Dance movement psychotherapist  1986-03-28   

                          trans_num   unix_time  merch_lat  merch_long  \
0  0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704 -112.154481   
3  6b849c168bdad6f867558c3793159a81  1325376076  47.034331 -112.561071   
4  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999  -78.632459   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1296675 entries, 0 to 1296674
Data columns (total 23 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   Unnamed: 0             1296675 non-null  int64  
 1   trans_date_trans_time  1296675 non-null  object 
 2   cc_num                 1296675 non-null  int64  
 3   merchant               1296675 non-null  object 
 4   category               1296675 non-null  object 
 5   amt                    1296675 non-null  float64
 6   first                  1296675 non-null  object 
 7   last                   1296675 non-null  object 
 8   gender                 1296675 non-null  object 
 9   street                 1296675 non-null  object 
 10  city                   1296675 non-null  object 
 11  state                  1296675 non-null  object 
 12  zip                    1296675 non-null  int64  
 13  lat                    1296675 non-null  float64
 14  long              

In [8]:
# Unique values of each column
for col in data.columns:
  print(f'{col}: {data[col].nunique()}')

Unnamed: 0: 1296675
trans_date_trans_time: 1274791
cc_num: 983
merchant: 693
category: 14
amt: 52928
first: 352
last: 481
gender: 2
street: 983
city: 894
state: 51
zip: 970
lat: 968
long: 969
city_pop: 879
job: 494
dob: 968
trans_num: 1296675
unix_time: 1274823
merch_lat: 1247805
merch_long: 1275745
is_fraud: 2


In [9]:
# Picking features
df = data[['merchant','category','amt','gender','zip','state','city_pop','job','is_fraud']]
df


merchant       category     amt gender  \
0                 fraud_Rippin, Kub and Mann       misc_net    4.97      F   
1            fraud_Heller, Gutmann and Zieme    grocery_pos  107.23      F   
2                       fraud_Lind-Buckridge  entertainment  220.11      M   
3         fraud_Kutch, Hermiston and Farrell  gas_transport   45.00      M   
4                        fraud_Keeling-Crist       misc_pos   41.96      M   
...                                      ...            ...     ...    ...   
1296670                    fraud_Reichel Inc  entertainment   15.56      M   
1296671             fraud_Abernathy and Sons    food_dining   51.70      M   
1296672                 fraud_Stiedemann Ltd    food_dining  105.93      M   
1296673  fraud_Reinger, Weissnat and Strosin    food_dining   74.90      M   
1296674  fraud_Langosh, Wintheiser and Hyatt    food_dining    4.30      M   

           zip state  city_pop                                job  is_fraud  
0        28654    NC      3495          Psychologist, counselling         0  
1        99160    WA       149  Special educational needs teacher         0  
2        83252    ID      4154        Nature conservation officer         0  
3        59632    MT      1939                    Patent attorney         0  
4        24433    VA        99     Dance movement psychotherapist         0  
...        ...   ...       ...                                ...       ...  
1296670  84735    UT       258                       Geoscientist         0  
1296671  21790    MD       100   Production assistant, television         0  
1296672  88325    NM       899                    Naval architect         0  
1296673  57756    SD      1126              Volunteer coordinator         0  
1296674  59871    MT       218           Therapist, horticultural         0  

[1296675 rows x 9 columns]

## One hot encoding features

Data will need to be in numerical form to be used in a ML model so we will experiment with different

In [11]:
# Convert categorical data into numerical using integer encoding
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
for col in df.select_dtypes(include='object'):
  df[col] = le.fit_transform(df[col])

df

C:\Users\tulan\AppData\Local\Temp\ipykernel_17796\2125786833.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = le.fit_transform(df[col])
C:\Users\tulan\AppData\Local\Temp\ipykernel_17796\2125786833.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = le.fit_transform(df[col])
C:\Users\tulan\AppData\Local\Temp\ipykernel_17796\2125786833.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

merchant  category     amt  gender    zip  state  city_pop  job  \
0             514         8    4.97       0  28654     27      3495  370   
1             241         4  107.23       0  99160     47       149  428   
2             390         0  220.11       1  83252     13      4154  307   
3             360         2   45.00       1  59632     26      1939  328   
4             297         9   41.96       1  24433     45        99  116   
...           ...       ...     ...     ...    ...    ...       ...  ...   
1296670       499         0   15.56       1  84735     44       258  215   
1296671         2         1   51.70       1  21790     20       100  360   
1296672       599         1  105.93       1  88325     32       899  308   
1296673       509         1   74.90       1  57756     41      1126  485   
1296674       370         1    4.30       1  59871     26       218  467   

         is_fraud  
0               0  
1               0  
2               0  
3               0  
4               0  
...           ...  
1296670         0  
1296671         0  
1296672         0  
1296673         0  
1296674         0  

[1296675 rows x 9 columns]

In [12]:
df = data[['merchant','category','amt','gender','zip','state','city_pop','job','is_fraud']]

In [13]:
# One hot encode data
df2 = pd.get_dummies(data=df, columns=['merchant','category','gender','state','job'], dtype='int')
df2

amt    zip  city_pop  is_fraud  merchant_fraud_Abbott-Rogahn  \
0          4.97  28654      3495         0                             0   
1        107.23  99160       149         0                             0   
2        220.11  83252      4154         0                             0   
3         45.00  59632      1939         0                             0   
4         41.96  24433        99         0                             0   
...         ...    ...       ...       ...                           ...   
1296670   15.56  84735       258         0                             0   
1296671   51.70  21790       100         0                             0   
1296672  105.93  88325       899         0                             0   
1296673   74.90  57756      1126         0                             0   
1296674    4.30  59871       218         0                             0   

         merchant_fraud_Abbott-Steuber  merchant_fraud_Abernathy and Sons  \
0                                    0                                  0   
1                                    0                                  0   
2                                    0                                  0   
3                                    0                                  0   
4                                    0                                  0   
...                                ...                                ...   
1296670                              0                                  0   
1296671                              0                                  1   
1296672                              0                                  0   
1296673                              0                                  0   
1296674                              0                                  0   

         merchant_fraud_Abshire PLC  \
0                                 0   
1                                 0   
2                                 0   
3                                 0   
4                                 0   
...                             ...   
1296670                           0   
1296671                           0   
1296672                           0   
1296673                           0   
1296674                           0   

         merchant_fraud_Adams, Kovacek and Kuhlman  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   
...                                            ...   
1296670                                          0   
1296671                                          0   
1296672                                          0   
1296673                                          0   
1296674                                          0   

         merchant_fraud_Adams-Barrows  ...  job_Visual merchandiser  \
0                                   0  ...                        0   
1                                   0  ...                        0   
2                                   0  ...                        0   
3                                   0  ...                        0   
4                                   0  ...                        0   
...                               ...  ...                      ...   
1296670                             0  ...                        0   
1296671                             0  ...                        0   
1296672                             0  ...                        0   
1296673                             0  ...                        0   
1296674                             0  ...                        0   

         job_Volunteer coordinator  job_Warden/ranger  job_Warehouse manager  \
0                                0                  0                      0   
1                                0         

Though the 2nd one is way longer, it will likely be the more useful dataset so we'll go with that one. Keeping track of what 100+ integers mean will be too confusing and time consuming.

The data at this point should be nearly ready for ingestion into a ML model to hopefully make some predictions.

# Project: Milestone 4 Build Your Model



Now that the data is properly prepared, we can begin to build the model up 

In [18]:
df = df2
df.head()

amt    zip  city_pop  is_fraud  merchant_fraud_Abbott-Rogahn  \
0    4.97  28654      3495         0                             0   
1  107.23  99160       149         0                             0   
2  220.11  83252      4154         0                             0   
3   45.00  59632      1939         0                             0   
4   41.96  24433        99         0                             0   

   merchant_fraud_Abbott-Steuber  merchant_fraud_Abernathy and Sons  \
0                              0                                  0   
1                              0                                  0   
2                              0                                  0   
3                              0                                  0   
4                              0                                  0   

   merchant_fraud_Abshire PLC  merchant_fraud_Adams, Kovacek and Kuhlman  \
0                           0                                          0   
1                           0                                          0   
2                           0                                          0   
3                           0                                          0   
4                           0                                          0   

   merchant_fraud_Adams-Barrows  ...  job_Visual merchandiser  \
0                             0  ...                        0   
1                             0  ...                        0   
2                             0  ...                        0   
3                             0  ...                        0   
4                             0  ...                        0   

   job_Volunteer coordinator  job_Warden/ranger  job_Warehouse manager  \
0                          0                  0                      0   
1                          0                  0                      0   
2                          0                  0                      0   
3                          0                  0                      0   
4                          0                  0                      0   

   job_Waste management officer  job_Water engineer  \
0                             0                   0   
1                             0                   0   
2                             0                   0   
3                             0                   0   
4                             0                   0   

   job_Water quality scientist  job_Web designer  job_Wellsite geologist  \
0                            0                 0                       0   
1                            0                 0                       0   
2                            0                 0                       0   
3                            0                 0                       0   
4                            0                 0                       0   

   job_Writer  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 1258 columns]

## Preparing the data for the Model

In [20]:
from sklearn.model_selection import train_test_split

1296675 rows is probably overkill, so we first need to reduce the amount of rows for this project.

In [22]:
df_RF = df.sample(n=100000, random_state=42)

In [23]:
df_RF['is_fraud'].value_counts()

is_fraud
0    99402
1      598
Name: count, dtype: int64

In [24]:
# Split the data
X = df_RF.drop('is_fraud', axis=1)
y = df_RF['is_fraud']

In [25]:
# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

## Preparing the Model

Now that the data is prepared, we will now need to choose the model for the data. These are the top models I am considering:
- Random Forest: A solid choice for many ML problems
- Neural Networks: Complex, highly accurate model
- XGBoost: Ensemble learning, good for large datasets

## Random Forest 

I decided to go with the Random Forest model for now, as it's a well balanced model that can produce good results.

In [28]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

In [29]:
# Initialize the Random Forest model
random_forest_model = RandomForestClassifier(n_estimators=50, random_state=42)

# Train the model on the training data
random_forest_model.fit(X_train, y_train)

RandomForestClassifier(n_estimators=50, random_state=42)

In [30]:
# Make predictions on the testing data
y_pred = random_forest_model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Get a classification report
print(classification_report(y_test, y_pred))

Accuracy: 0.9956
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     24859
           1       0.92      0.24      0.38       141

    accuracy                           1.00     25000
   macro avg       0.96      0.62      0.69     25000
weighted avg       1.00      1.00      0.99     25000



The model performed poorly, but as we can see from the support column, the amount of columns I picked with actual fraud data was very little, so we're going to edit the data to choose more columns that were in fact fraudulent.

In [32]:
df['is_fraud'].value_counts()

is_fraud
0    1289169
1       7506
Name: count, dtype: int64

There's only around 7500 fraudelent columns total so we're going to ensure that all of those get picked, as well as about 50,000 non-fraud columns, since 100,000 looked to be overkill here. 

In [34]:
# Selecting all columns where 'is_fraud' equals 1
fraudulent_data = df[df['is_fraud'] == 1]

# Selecting 25,000 random columns where 'is_fraud' equals 0
non_fraudulent_data = df[df['is_fraud'] == 0].sample(n=25000, random_state=42)

In [35]:
combined_data = pd.concat([fraudulent_data, non_fraudulent_data])
combined_data

amt    zip  city_pop  is_fraud  merchant_fraud_Abbott-Rogahn  \
2449     281.06  28611       885         1                             0   
2472      11.52  78208   1595797         1                             0   
2523     276.31  78208   1595797         1                             0   
2546       7.03  28611       885         1                             0   
2553     275.73  78208   1595797         1                             0   
...         ...    ...       ...       ...                           ...   
924495     4.84  27258      6006         0                             0   
1135644    5.15  78248   1595797         0                             0   
843129   131.69  98238        85         0                             0   
855862    54.24  18433      6508         0                             0   
665512    73.38  36532     27829         0                             0   

         merchant_fraud_Abbott-Steuber  merchant_fraud_Abernathy and Sons  \
2449                                 0                                  0   
2472                                 0                                  0   
2523                                 0                                  0   
2546                                 0                                  0   
2553                                 0                                  0   
...                                ...                                ...   
924495                               0                                  0   
1135644                              0                                  0   
843129                               0                                  0   
855862                               0                                  0   
665512                               0                                  0   

         merchant_fraud_Abshire PLC  \
2449                              0   
2472                              0   
2523                              0   
2546                              0   
2553                              0   
...                             ...   
924495                            0   
1135644                           0   
843129                            0   
855862                            0   
665512                            0   

         merchant_fraud_Adams, Kovacek and Kuhlman  \
2449                                             0   
2472                                             0   
2523                                             0   
2546                                             0   
2553                                             0   
...                                            ...   
924495                                           0   
1135644                                          0   
843129                                           0   
855862                                           0   
665512                                           0   

         merchant_fraud_Adams-Barrows  ...  job_Visual merchandiser  \
2449                                0  ...                        0   
2472                                0  ...                        0   
2523                                0  ...                        0   
2546                                0  ...                        0   
2553                                0  ...                        0   
...                               ...  ...                      ...   
924495                              0  ...                        0   
1135644                             0  ...                        0   
843129                              0  ...                        0   
855862                              0  ...                        0   
665512                              0  ...                        0   

         job_Volunteer coordinator  job_Warden/ranger  job_Warehouse manager  \
2449                             0                  0                      0   
2472                             0         

Now let's try again.

In [37]:
df_RF = combined_data

In [38]:
df_RF['is_fraud'].value_counts()

is_fraud
0    25000
1     7506
Name: count, dtype: int64

In [39]:
# Split the data
X = df_RF.drop('is_fraud', axis=1)
y = df_RF['is_fraud']

In [40]:
# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [41]:
# Initialize the Random Forest model
random_forest_model = RandomForestClassifier(n_estimators=50, random_state=42)

# Train the model on the training data
random_forest_model.fit(X_train, y_train)

RandomForestClassifier(n_estimators=50, random_state=42)

In [42]:
# Make predictions on the testing data
y_pred = random_forest_model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Get a classification report
print(classification_report(y_test, y_pred))

Accuracy: 0.964316475944383
              precision    recall  f1-score   support

           0       0.96      0.99      0.98      6260
           1       0.98      0.87      0.92      1867

    accuracy                           0.96      8127
   macro avg       0.97      0.93      0.95      8127
weighted avg       0.96      0.96      0.96      8127



Even though the accuracy of the model decreased slightly, it did a MUCH better job on the columns that were actually fraudulent. 

The model's weakest point was it's recall on fraudelent purchases. This is a concern because that is the most essential thing to know when it comes to fraudulent activity. A missed positive means lost revenue. The high accuracy on non-fraud purchases is important too, preventing a reduction in customer satisfaction with false accusations.

## Hyperparameter Tuning

The Random Forest Classifer performed well so we will now tune the model to try and increase the overall accuracy. 

In [45]:
from sklearn.model_selection import GridSearchCV

In [46]:
# Define hyperparameter grid
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [5, 10, 15],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

(Initally included "none" in max_depth but that would make this take too long.)

In [48]:
# Create Grid Search object
grid_search = GridSearchCV(estimator=random_forest_model, param_grid=param_grid, cv=5, scoring='accuracy')

# Fit the grid search to the data
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=RandomForestClassifier(n_estimators=50, random_state=42),
             param_grid={'max_depth': [5, 10, 15],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 200, 300]},
             scoring='accuracy')

In [49]:
# Get the best model
best_params = grid_search.best_params_
best_params

{'max_depth': 15,
 'min_samples_leaf': 1,
 'min_samples_split': 10,
 'n_estimators': 300}

In [55]:
best_model = grid_search.best_estimator_

In [66]:
best_model

RandomForestClassifier(max_depth=15, min_samples_split=10, n_estimators=300,
                       random_state=42)

In [59]:
# Make predictions on the test set
y_pred = best_model.predict(X_test)

In [61]:
# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Get a classification report
print(classification_report(y_test, y_pred))

Accuracy: 0.8947951273532669
              precision    recall  f1-score   support

           0       0.88      1.00      0.94      6260
           1       1.00      0.54      0.70      1867

    accuracy                           0.89      8127
   macro avg       0.94      0.77      0.82      8127
weighted avg       0.91      0.89      0.88      8127



## Model Evaluation

As we can see from the results, tuning the parameters helped the model slightly, but not by much. Higher depth and estimators did make a positive impact though. 96.5% accuracy for a model is very good overall.

Max depth is probably the most impactful of the parameters and would probably be the best thing to tweak, then estimators. 

## Conclusion

Overall I was pleased with the results. There were a lot of features and a decent amount of rows so the model was given an adequate amount of information.

A potential next step would be trying a different model and seeing if it performs better in the recall section, or just overall, but I'd say this model was a success. Tweaking the hyperparameters don't seem overly neccesary since they didn't increase the accuracy much, but it is good to experiment. 

In [68]:
import joblib

In [173]:
# Save the model to a file
joblib.dump(best_model, 'best_RFC_model_fraud.joblib')

['best_RFC_model_fraud.joblib']

# Project: Milestone 5 Create Your Application to Finish the Project

Use Dash to create an application that uses your model. You will follow a process similar to the one shown in the text (except with your model and your data elements).

In [74]:
# pip install dash

In [157]:
import dash
import pickle
from dash import dcc
from dash import html, State, callback
from dash.dependencies import Input, Output
from sklearn.preprocessing import StandardScaler


In [183]:
# Load your model

model = joblib.load('best_RFC_model_fraud.joblib')
model.fit(X_train, y_train)

RandomForestClassifier(max_depth=15, min_samples_split=10, n_estimators=300,
                       random_state=42)

In [185]:
# Reloading the data from before
data = combined_data


## Creating the Dash app part 1

Here I tried to re-create the most releveant looking dash application from the Dash documentation:

In [187]:
# Process to create the dash app, including pre-processing, the layout, callback, and output.

def preprocess_features(features):
    # Convert features to appropriate data types
    amount = float(features[0])
    zip_code = str(features[1])
    city_population = int(features[2])
    processed_features = [amount, zip_code, city_population]

    return processed_features

def predict(amount, zip_code, city_population):
    features = [amount, zip_code, city_population]
    processed_features = preprocess_features(features)
    prediction = model.predict([processed_features])[0]  # Extract prediction from array
    return prediction

app = dash.Dash(__name__)

app.layout = html.Div([
    dcc.Input(id='transaction-amount', type='number', placeholder='Transaction Amount'),
    dcc.Input(id='transaction-zip', type='text', placeholder='Transaction Zip'),
    dcc.Input(id='transaction-city-population', type='number', placeholder='City Population'),
    # I chose these 3 as I felt they were the most relevant and easiest for someone to input as they are already numbers.
    html.Button('Predict', id='submit-button'),
    html.Div(id='prediction-output')
])

@app.callback(
    Output('prediction-output', 'children'),
    [Input('submit-button', 'n_clicks')],
    [State('transaction-amount', 'value'),
     State('transaction-zip', 'value'),
     State('transaction-city-population', 'value')]
)
def update_output(n_clicks, amount, zip_code, city_population):
    if n_clicks is None:
        return 'No prediction yet'

    prediction = predict(amount, zip_code, city_population)
    return f'The transaction is predicted to be fraudulent: {prediction}'

if __name__ == '__main__':
    app.run_server(debug=True)

Unfortunately it seems that the dash app wants an input for every column, but I have about 1257 input columns.

I looked into the matter but I couldn't figure out how to fix it as I don't know Dash all to well yet.

## Creating the Dash app part 2

For the sake of time i am going to make a sample app with just 3 columns just to see if it'll work.

In [195]:
# Select relevant features
X = data[['amt', 'zip', 'city_pop']]
y = data['is_fraud']

# Create and train the model
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(max_depth=15, min_samples_split=10, n_estimators=300,
                       random_state=42)

In [197]:
# Process to create the dash app, including pre-processing, the layout, callback, and output. (Same as above)
def predict(amount, zip_code, city_population):
    features = [amount, zip_code, city_population]
    prediction = model.predict([features])[0]
    return prediction

app = dash.Dash(__name__)

app.layout = html.Div([
    dcc.Input(id='transaction-amount', type='number', placeholder='Transaction Amount'),
    dcc.Input(id='transaction-zip', type='text', placeholder='Transaction Zip'),
    dcc.Input(id='transaction-city-population', type='number', placeholder='City Population'),
    html.Button('Predict', id='submit-button'),
    html.Div(id='prediction-output')
])

@app.callback(
    Output('prediction-output', 'children'),
    [Input('submit-button', 'n_clicks')],
    [State('transaction-amount', 'value'),
     State('transaction-zip', 'value'),
     State('transaction-city-population', 'value')]
)
def update_output(n_clicks, amount, zip_code, city_population):
    if n_clicks is None:
        return 'No prediction yet'

    prediction = predict(amount, zip_code, city_population)
    return f'The transaction is predicted to be fraudulent: {prediction}'

if __name__ == '__main__':
    app.run_server(debug=True)

Looks like the app works and can make predictions now. If I had more time my next steps would be:
- Identifying more features to add to the app
- Potentially finding a way to make the app only need a few inputs, but consider the data as a whole (not sure if this is possible or not)
- Try different models (maybe different models are more compatible with Dash as a whole? Can't say.)
- Do some more edits to the data (not sure if this would make a difference but worth a shot

## Conclusion

In conclusion, this type of use with Dash could help generate data and predictions for fraudulent purchases and what they might look like. how that looks in practice I can't say for sure, but the application and practice is surely there.